### PR TITLE
KVS refactor of lookup() / walk() functions

### PR DESCRIPTION
**Pull request by chu11**
In this PR, we convert the KVS walk function from a recursive implementation into a non-recursive one and simplified the logic as well.  The predominant logic changes are:

A) In preparation for maintaining "state" so that RPCs aren't fully replayed, the entire path of the kvs get request is pre-parsed and stored in a list.

B) The walk() function has been made non-recursive and "depth" is maintained in a stack.  This is also in preparation for maintaining "state" for removal of RPC replay.

C) A variety of	code logic changes to make the walk() function simpler.

The predominant logic changes that may look strange in this PR are:

A) The pre-parsing of the path makes things more inefficient for the time being b/c RPCs are still completely replayed (i.e. it is fully parsed on every RPC-then continuation).

B) There are some paths (such as through get_request_cb()) in which a load() will be called twice on the same root reference.  This will be resolved when lookup() is refactored.  This PR is only sticking to walk().

I still need to run through valgrind to make sure I didn't mem-leak anything.  But wanted to throw the PR up here for a first skim.

### CODE REVIEW

**Comment by garlick** [comment]
Great job @chu11 -  this was complicated, messy, and critical code and now it is definitely heading in the right direction.

As we discussed privately, I think clarity (and my ability to review these changes) would improve significantly if `walk()` could be broken out to a separate C module with its public and private bits clearly separated and its purpose and function documented.  Ideally for the history, the individual commits to move the function to a new file, and the changes to the function would be kept separate, although if that proves too difficult, I'd be willing to sacrifice the diff(s) between old and new walk functions in the history, since it's pretty hard to follow anyway.

Note the work on "futures" in #1053.  Possibly walk or some future more reusable incarnation could return a `flux_future_t` that encapsulates its state and the ability to call it in both synchronous and asynchronous modes.  That's just something to keep in the back of your mind, not concrete feedback for this PR.

**Comment by chu11** [comment]
@garlick It was a little hard to splice out b/c of the dependencies (such as the ```load()``` function).  In my mind I was hoping to do the file split later, but perhaps it might be better to do that now as a checkpoint.

Thinking about it, I think I see a way to slice things out now that might be pretty good.  As you mentioned privately, ```load()``` can probably be migrated into the cache API.  Finally de-coupling that from the main kvs file.

**Comment by garlick** [comment]
If it can be done in the near term it would probably be good to do it in this PR.  If it's a really big project though, then maybe we should consider putting this in as is.

**Comment by chu11** [comment]
My idea to splice things out didn't work out.  A lot of it centers on the fact that the ```load()``` function performs cache operations, rpcs, and waits.  So it's sort of has a dependency on everything.

I sort of thought this was a good "checkpoint" for a PR, but perhaps its not.  My general idea was:

1. refactor ```walk()``` w/ state
2. refactor ```lookup()``` w/ state
3. refactor ```get_request_cb()``` to utilize state
4. refactor ```get_request_cb()``` to call ```load()``` instead of ```lookup()``` and ```walk()```.  The latter functions now call ```cache_lookup()```, and on failure pass back a reference for ```get_request_cb()``` to call load with.
5. splice out ```lookup()``` and ```walk()``` into new file, as they only call cache API and no longer deal with RPCs.

The order was predominantly based on the fact that w/ "state", it would be easy to pass back information to ```get_request_cb()```.  However, 4 & 5 could be done before 1 & 2 & 3, just figured it would be uglier.  Lemme see if that'll work out easier/harder than I thought.

**Comment by chu11** [comment]
It ended up doing step 4 above wasn't as difficult as I thought it'd be.  Perhaps the code isn't the prettiest, but it's not bad.  It'll allow us to splice off the ```walk()``` function into a separate API in this PR.  

I'll issue a new PR for the ```load()``` refactoring.

**Comment by garlick** [comment]
@chu11 sorry for not providing more review, wasn't sure if I was waiting for more or not. 
  Superficially it looks good to me but I don't feel confident that I'm understanding/remembering every corner case.  I do think we have decent test coverage for this so the fact that tests are passing is a good indication.

Were you able to verify that memory isn't leaking?

Another thing to check just to be sure we're not causing problems for others would be the soak test.   We could get a baseline that could serve all the KVS rework activity and then run it for successive PR's.  (Some past results are posted to various closed issues - e.g. #614)

I'm running an errand and when I return around 2 I'll have one more spin through the code.

**Comment by chu11** [comment]
@garlick Yeah, I ran a bunch of valgrind runs using the ```src/test/valgrind.sh```.  Also wrote an alternate cheapo workload consisting of a collection of simple kvs operations.  The leak summary at the very end didn't change, which I think is the most important thing to verify didn't change.

I'm presently re-building this tree to be rebased under the patches from PR #1069.  With PR #1069, I can splice ```lookup()``` and ```walk()``` into a new file.  So hopefully this PR will be easier to understand, as all the changes will be more isolated.

**Comment by garlick** [comment]
That sounds perfect, thanks.

**Comment by chu11** [comment]
Pushed an update, based on the branch from #1069 .  With that branch, I can now split out the ```lookup()``` and ```walk()``` functions into its own file.  The nice thing is now ```lookup()``` is the only public facing function.  Everything behind that is now private (```walk()``` as well as my collection of helper functions and helper structs).

The bad thing is that with ```lookup()``` being the "public" function of this file, it looks sorta bad given I haven't completed the lookup-refactor.  Which sort of makes this branch look sad and incomplete.  Although I was hoping to "checkpoint" with refactoring just ```walk()```, perhaps I'll have to do the rest in this PR.  Which will ultimately end up being a really large patch series.

**Comment by garlick** [comment]
I think it'd be OK to continue on with the refactoring in this PR.  After all it's a pretty focused topic.

Wow, it's really nice having that split off.  I wonder if we can do anything about the appalling number of in/out parameters in this function?  That might bear some pondering.  Maybe we could turn it into more of a class - for example, maybe
```c
struct lookup *lookup_create (struct cache *cache);
int lookup_set_epoch (struct lookup *l, int epoch); // call from heartbeat_cb
...
```

seems like root, rootdir, and root_dirent should be packaged up in a struct?

maybe the result (alp, load_ref, stall, ep) could also be packaged up?

(I'm probably not seeing all the ramifications of what I'm suggesting, but it would be nice if that parameter list was a little easier to look at :-)

Great work.

**Comment by garlick** [comment]
I meant to add: unit test?  `cache.c` has one so maybe it would not be too far fetched for `lookup.c` since the cache is its main dependency?  If you're going to be doing more refactoring you may find it is worth adding one sooner rather than later?

**Comment by chu11** [comment]
@garlick Yeah, the appalling number of parameters is something I'm working on.  In a prior branch I had a ```struct lookup``` that I passed around, but now that it's split off into it's own file, I was thinking it'll be something like you're suggesting.

Unit tests were something I was going to do after the refactor of ```lookup()```.  I don't even know what the final API is even going to look like, so I decided it would be best for later.

I think PR #1069 is about the best mini-checkpoint on this refactor.  If we're going to go full steam ahead w/ a mega PR here, should we just drop PR #1069?

**Comment by garlick** [comment]
I'm OK with the mega PR approach, especially if we end up taking a few wrong turns and can squash that down in the history.  If this PR is a superset of that one, and you agree with the approach, then yeah I guess that PR could be abandoned.

**Comment by chu11** [comment]
Finished up something that looks near complete yesterday evening.  Still need to valgrind and soak it and add unit tests.  The key completion is that now ```get_request_cb``` and ```watch_r1equest_cb``` no longer replay rpcs fully, they maintain state and pickup things in the middle.  ```lookup()``` is now API-ized.

I think I can splice out a few cleanup and "setup convenience functions" PRs out of this to shorten the number of commits.

**Comment by chu11** [comment]
Just pushed a branch w/ lookup unit tests.  Whew, there's a lot of them.

Running valgrind, no mem-leaks based on the leak summary staying the same.  Based on ```valgrind.sh``` about 2.3% more allocations and around 1.2% more memory allocated during the workload.  This isn't too surprising, as maintaining "state" in kvs lookups will lead to more allocations.  I was actually concerned it was going to be much higher, so this isn't that bad.  Tweaked it down best I could with some memory allocation games.  Still gotta run soak tests.

So when replaying RPCs, there is one gotcha, which is how do you know if you are replaying a RPC or if it's the original RPC being run the first time?  Because if we are maintaining state during replay of RPCs, we need to pass this state information to the replay.

A) In a play-branch, I originally replayed to a new function, so it was obvious it was a replay and data passed to it was clearly state data.  I didn't like this, as it caused more code duplication than I'd like.

B) In an earlier version of this branch, I set some auxiliary data structures in the flux message but it required to creating a new ```wait_msg_aux_set()``` function.  Don't like that ```wait_msg_aux_set()``` function, b/c it exposes what should be abstracted.  In addition, setting auxiliary messages adds some more mallocs to every lookup (internal zhash in message).

C) In this branch, I simply pass the auxiliary data structure as the "arg" on the next replay.  Using magic numbers (i.e. ```ctx->magic == KVS_MAGIC```), determine what data was passed in and use that to determine if its the initial RPC or a replay.  I dislike the idea of checking for data types this way, but probably is the best way and minimizes extra mallocs.

**Comment by chu11** [comment]
huh, hit ```FAIL: t0016-cron-faketime.t 6 - flux-cron tab works for day of month``` .  Restarting travis.

**Comment by garlick** [comment]
Hey @chu11, I'll plan to run the soak tests, both baseline and on this branch, after lunch.  This is looking great so far.

**Comment by chu11** [comment]
I just realized a way to eliminate the ```walk_pathcomp_t``` struct completely.  Just one of those things when you see things a certain way at first.

Test hacks eliminated a large number of mallocs.  Under a valgrind test, there were fewer total allocations than even master, but a bit more memory allocated total still.

master
```
total heap usage: 5,343,188 allocs, 5,342,949 frees, 422,140,268 bytes allocated
```

my new branch
```
total heap usage: 5,337,052 allocs, 5,336,813 frees, 424,614,026 bytes allocated
```

Soak gave me a slightly worse average on one run compared to master (which I couldn't believe), so I ran again and got a faster average than master.  I suppose some normal variance due to whos on the machine.  I'll pretty up the branch and re-push.

**Comment by garlick** [comment]
In `walk()`, `zlist_remove` is called within a walk of the list.  I was going to call this out as unsafe but it turns out that with commit cf1f4850750e303118f0e2fe376d2c71db6a92a0 (in 2013, in v2.0.0!) czmq addressed this and it is now safe.

**Comment by garlick** [comment]
The refactoring and isolation of `lookup()` is great.  The extra work avoided by not starting from the root at each stall/restart is great.   The new tests are great.  We've verified with valgrind and the soak test that these changes don't introduce any obvious regressions.  Still a lot of KVS refactoring needed, but this step forward makes it all seem more tractable.

I think we should go ahead and get this merged.  What do you think @chu11?  Ready?

**Comment by chu11** [comment]
@garlick Close, trying to eek out a few more performance tweaks (removed an uncommon branch check), used a switch statement on lookup "states" which I think is cleaner and might be faster too.

**Comment by chu11** [comment]
@garlick From the point of your last comment, converted the if-state-branch logic in ```lookup()``` to be a switch, as I think it makes the code easier to understand.  Also added more stall unit tests as I realized I didn't do the stall unit tests correctly (I was re-starting the lookups from the beginning again, which is not the right idea).  A few minor cleanups.  But other than that, I think the branch is good to go.

**Comment by garlick** [comment]
OK I think it's good to go once travis passes.
